README.md
# What it does
Upgrade PaloAlto firewalls from 8.1.xx to 9.1.xx with PowerShell.  The firewalls will need DNS and general internet access to proceed.  It will install device licesnses.  Then it will install your requested PanOS version.  Afterwards it will install the latest App/Threats/AV Dynamic updates.  Then it will move into the next IP in the list to upgrade.

# Technologies
* Powershell 5
* PaloAlto Networks PanOS 8.1/9.1

# Recommended uses 
Rapid deployment of of firewalls to get them to a common level of PanOS.

# How to use
1) Edit firewallstoupgrade.txt with 1 IP per line
2) Run the script and enter the version you want to upgrade to
3) Enter username and password (or hard code it with default username and password)
4) Enter Y/N for the reboot of the firewall after install for saftey reasons (or remove this condition)
5) Do something else while your firewalls upgrade


# If you want to get froggy (at your own risk)
Comment lines 113-114 and remove the "}" on line 124 for set it and forget it mode.

```powershell
$confirm = read-host "DO YOU WANT TO REBOOT $ip? Y/N?"
if ($confirm -eq "Y" -or $confirm -eq "y") { 
```


changelog
# 7/30/2020 bk
# Fixed an issue that caused it not to install PanOS
# Changed the if statement that checks if it's ready to reboot to the install PanOS result
# Fixed some ordering issues with the "isjobdone" function

# 7/73/2020 bk
# Converted loops that check for job completion into function
# test branch

# 7/73/2020 bk
# Fixed various bugs with the loops
# Added apps/threats and av install after upgrade

# change log 4-22-2020 bk
# Fixed formatting
# Moved the ignore self signed cert into "ignore self signed certificate.ps1"
# Changed firewall list for upgrading into a variable to use present directory
#   still broke fix it later
#   update readme to reflect this change

#Loop to check on PanOS downloading #update to the install method probably won't quit loop *fixed loop result was not FIN but OK*
#Changed version to variable
#$stoploop = $result.SelectSingleNode("//job") | Select-Object status, result # removed this *broke loop*
#   fixed this loop also
#line 115 erroraction silentcontinue doesn't work right
#tested upgrade from 9.1.0 to 9.1.2 seems to be compatable still - need to verify later
#Updated readme and removed extra comments

test.ps1
# Palo Alto API base config upgrade 8.1.x or 9.1.x
# Version: Aplha-0.5 7/23/2020
# Written by Brandon Kelley

# Edit firewallstoupgrade.txt with 1 IP per line
# Currently there's manual Y/N for the reboot of the firewall after install for saftey reasons
# At this time.

# Future - convert the API calls into variables so it's more modular and less code
# Add more error handling and input validation
# Create more functions, so debugging will be easier
# Learn to create parallell proccessing to do multiple firewalls at the same time
# add timer for rebooting.

#Ignore self signed certificate calling this script in same dir
$ignorecert = "$PSScriptRoot\ignore self signed certificate.ps1"
&$ignorecert

#text file for IPs you want to configure
$firewalList = (get-content "$PSScriptRoot\firewallstoupgrade.txt")

#ask user for what version of code you want *need to add 2-3 code jump process via function*
$version = read-host 'version you want to go to'

function getapikey {
    # Get API Key with username and pw prompts
    param ([string]$firewallip)
    [xml]$content = Invoke-RestMethod -uri "https://$firewallip/api/?type=keygen&user=$user&password=$password"
    $content.response.result.key
}

#testing job loop function
function isjobdone {
    [cmdletbinding()]param ([string]$job, [string]$ip, [string]$key, [string]$whatareyoudoing)
    $loop = $null
    while ($loop.result -ne "OK" -and $loop.status -ne "FIN") {
        [xml]$amidone = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
        write-host $whatareyoudoing $amidone.response.result.job.progress "%"
        if($amidone.response.msg.line -ne $null){
            write-host $amidone.response.msg.line
            $loop.result = "OK" #needs testing
            $loop.status = "FIN"
        }
    $loop = $amidone.response.result.job | Select-Object status, result
    if ($loop.result -eq "OK" -and $loop.status -eq "FIN") { write-host "$job job done"}
    start-sleep -s 15
    }    
}

#if there's no key hash table it'll run from here
$keys = $null
if ($keys.values -eq $null) {
    #create keys hash table
    $keys = @{ }
    #get creds
    $user = read-host 'username'
    $password = read-host 'password'
    foreach ($firewall in $firewalList) {
        #temp is to run the api function to get the key per firewall
        $temp = getapikey($firewall)
        # Add the firewall IP as the first key in hash table and the API key as a value
        $keys.Add($firewall, $temp)
    }
    #removing passwords from memory
    $user = $null
    $password = $null
}
Measure-Command {
    # Run through the $keys hash table with getenumerator
    foreach ($item in $keys.GetEnumerator()) {
        #assign the key (IP) to individual variable
        $ip = $item.Key
        #assign the hash table key's value that has the API key to individual variable
        $key = $item.Value
        #invoke license fetch using variables from above
        [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><license><fetch></fetch></license></request>&key=$key"
        $result.response.status
        if ($result.response.status -ne "success") { write-host "failed license fetch on $ip" }
        # if the response isn't success write to screen - expand on this to provide error message or report after it runes
        #Check to see if PanOS is up to date
        [xml]$resultsysteminfo = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key"
        $cursoftware = $resultsysteminfo.SelectSingleNode("//sw-version")
        if ($cursoftware.'#text' -ne "$version") {
            #Fetch available software from internet
            [xml]$resultsoft = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><check></check></software></system></request>&key=$key"
            if ($resultsoft.response.status -ne "success") { write-host "failed on Fetch available software $ip" }
            $areYouUpgraded = $resultsoft.SelectSingleNode("//entry[version = ""$version""]") | Select-Object downloaded
            if ($areYouUpgraded.downloaded -eq 'no') {
                #download $version pan os 
                [xml]$resultdl = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><download><version>$version</version></download></software></system></request>&key=$key"
                $job = $resultdl.response.result.job
                isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Downloading PanOS"            
            }
            Else {
                # install $version that's downloaded
                [xml]$resultinstall = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><install><version>$version</version></install></software></system></request>&key=$key"
                $job = $resultinstall.response.result.job
                isjobdone -job $job -ip $ip -key $key
                write-host "$ip has installed $version.  Moving to reboot"                
            }
        }
        #verify panOS installed
        [xml]$resultinstalled = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><check></check></software></system></request>&key=$key"
        #add another condition to check for uploaded also
        $areyouinstalled = $resultinstalled.SelectSingleNode("//entry[version = ""$version""]") | Select-Object current, downloaded
        if ($areyouinstalled.current -eq "no" -and $areyouinstalled.downloaded -eq 'yes') {
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            $confirm = read-host "DO YOU WANT TO REBOOT $ip? Y/N?"
            if ($confirm -eq "Y" -or $confirm -eq "y") {
                [xml]$resultreboot = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><restart><system></system></restart></request>&key=$key"
                $resultreboot
                while ($cursoftware -ne "$version") {
                    #Check to see if PanOS is up to date *erroraction isn't working right fix it*
                    [xml]$resultfin = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key" -ErrorAction SilentlyContinue
                    $cursoftware = $resultfin.SelectSingleNode("//sw-version")
                    write-host "$ip is still booting"
                    start-sleep -s 15
                    if ($cursoftware.'#text' -eq "$version" -and $result.response.result -ne "Command succeeded with no output") { write-host "$ip has been upgraded" }
                }
            }
        }
        #download apps/threats
        [xml]$resultapp = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><content><upgrade><download><latest/></download></upgrade></content></request></request>&key=$key"
        $job = $resultapp.response.result.job
        isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Download apps/threats"
        
        #install apps
        [xml]$resultappins = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><content><upgrade><install><version>latest</version></install></upgrade></content></request>&key=$key"
        $job = $resultappins.response.result.job
        isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Installing apps/threats"
        
        #download AV
        [xml]$resultavdl = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><anti-virus><upgrade><download><latest/></download></upgrade></anti-virus></request>&key=$key"
        $job = $resultavdl.response.result.job
        isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Downloading AV"
        
        #install av
        [xml]$resultavins = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><anti-virus><upgrade><download><latest/></download></upgrade></anti-virus></request>&key=$key"
        $job = $resultavins.response.result.job
        isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Installing AV"
    }
}
#Reset variables for next Upgrade
<#
    $cursoftware = $null
    $areyouinstalled = $null
    $areYouUpgraded = $null
    $job = $null
    $result = $null
    #>
#add error handling


isjobdone.ps1
<#$ignorecert = "$PSScriptRoot\ignore self signed certificate.ps1"
&$ignorecert
function getapikey {
    # Get API Key with username and pw prompts
    param ([string]$firewallip)
    #create keys hash table
    $keys = @{ }
    #get creds
    $user = read-host 'username'
    $password = read-host 'password'
    #removing passwords from memory
        
    [xml]$content = Invoke-RestMethod -uri "https://$firewallip/api/?type=keygen&user=$user&password=$password"
    $content.response.result.key
    $user = $null
    $password = $null
}
#>
function isjobdone {
    # Get API Key with username and pw prompts
    [cmdletbinding()]param ([string]$job, [string]$ip, [string]$key, [string]$whatareyoudoing)
    $loop = $null
    while ($loop.result -ne "OK" -and $loop.status -ne "FIN") {
        [xml]$amidone = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
        write-host $whatareyoudoing $amidone.response.result.job.progress "%"
    $loop = $amidone.response.result.job | Select-Object status, result
    if ($loop.result -eq "OK" -and $loop.status -eq "FIN") { write-host "$job job done"}
    start-sleep -s 15
    }    
}

<# testing
$ip = '192.168.0.10' 
$key = 'LUFRPT1SazNTTUFxQWxkTFVIaThnM0c1ejJzWXZqaG89T3lYZWducUR3dE9BeC8zSW9TUlF3dGNvc1pQL3pGcWZvSVQ1cW5qNnY5U1hScS9QU3NSRk5WRWdFOEpITENLaA=='

        #download AV
        [xml]$resultavdl = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><anti-virus><upgrade><download><latest/></download></upgrade></anti-virus></request>&key=$key"
        $job = $resultavdl.response.result.job
        try { isjobdone -job $job -ip $ip -key $key -whatareyoudoing 'Downloading AV'
        }
        catch { if($resultavdl.response.msg.line -ne $null){write-host $resultavdl.response.msg.line}}
             

#$key = getapikey -firewallip 192.168.0.10

#isjobdone -job '781' -ip '192.168.0.10' -key 'LUFRPT1SazNTTUFxQWxkTFVIaThnM0c1ejJzWXZqaG89T3lYZWducUR3dE9BeC8zSW9TUlF3dGNvc1pQL3pGcWZvSVQ1cW5qNnY5U1hScS9QU3NSRk5WRWdFOEpITENLaA=='


<#
$stoploop3 = $null
[xml]$resultapp = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><content><upgrade><download><latest/></download></upgrade></content></request></request>&key=$key"
$job = $resultapp.response.result.job
while ($stoploop3.result -ne "OK" -and $stoploop3.status -ne "FIN") {
    [xml]$resultapp = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
    $stoploop3 = $resultapp.response.result.job | Select-Object status, result
    if ($stoploop3.result -ne "OK" -and $stoploop3.status -ne "FIN") { write-host $resultapp.response.result.job.progress "% complete downloading app/threats" }
    if ($stoploop3.result -eq "OK" -and $stoploop3.status -eq "FIN") { write-host "downloaded apps" }
    start-sleep -s 15
#>
#>

ignore self signed certificate.ps1
#ignore self signed certificate
if (-not ([System.Management.Automation.PSTypeName]'ServerCertificateValidationCallback').Type) {
    $certCallback = @"
    using System;
    using System.Net;
    using System.Net.Security;
    using System.Security.Cryptography.X509Certificates;
    public class ServerCertificateValidationCallback
    {
        public static void Ignore()
        {
            if(ServicePointManager.ServerCertificateValidationCallback ==null)
            {
                ServicePointManager.ServerCertificateValidationCallback +=
                    delegate
                    (
                        Object obj,
                        X509Certificate certificate,
                        X509Chain chain,
                        SslPolicyErrors errors
                    )
                    {
                        return true;
                    };
            }
        }
    }
"@
    Add-Type $certCallback
}
[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12;
[ServerCertificateValidationCallback]::Ignore()
################################################# End - ignore self signed certificate #####################################


Palo-Alto-8.1-OS-upgrader-XML-api.ps1
<###### Working Alpha #######
Palo Alto API base config upgrade 8.1.x or 9.1.x
Version: Aplha-1.0 7/30/2020
Written by Brandon Kelley

######  How to use  ######
Edit firewallstoupgrade.txt with 1 IP per line
Run the script and enter the version you want to upgrade to
Enter username and password
Enter Y/N for the reboot of the firewall after install for saftey reasons (or remove this condition)
Do something else while your firewalls upgrade

###### Future improvements ######
Create parallel proccessing to do multiple firewalls at the same time
Input validation
Add function error streams 
Make it more modular
#>

#Ignore self signed certificate calling this script in same dir
$ignorecert = "$PSScriptRoot\ignore self signed certificate.ps1"
&$ignorecert

# Edit this text file for IPs you want to upgrade
$firewalList = (get-content "$PSScriptRoot\firewallstoupgrade.txt")
write-host "Upgrading these firewalls:"
$firewalList
# Ask what version of code you want # need to add 2-3 code jump
$version = read-host 'version you want to go to'

function getapikey {
    # Get API Key with username and pw prompts
    param ([string]$firewallip)
    [xml]$content = Invoke-RestMethod -uri "https://$firewallip/api/?type=keygen&user=$user&password=$password"
    $content.response.result.key
}

# Check if the job complete 
function isjobdone {
    [cmdletbinding()]param ([string]$job, [string]$ip, [string]$key, [string]$whatareyoudoing)
    $loop = $null
    while ($loop.result -ne "OK" -and $loop.status -ne "FIN") {
        [xml]$amidone = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
        write-host $whatareyoudoing $amidone.response.result.job.progress "%"
        $loop = $amidone.response.result.job | Select-Object status, result
        # Write out error message from API call
        if ($amidone.response.msg.line -ne $null) {
            write-host $amidone.response.msg.line
            $loop.result = "OK" #needs testing
            $loop.status = "FIN"
        }
        # Check if job has finished
        if ($loop.result -eq "OK" -and $loop.status -eq "FIN") { write-host "$job job done" }
        start-sleep -s 15
    }    
}

# If there's no keys hash table it'll run from here
$keys = $null
if ($keys.values -eq $null) {
    # Create hash table
    $keys = @{ }
    # Get creds
    $user = read-host 'username'
    $password = read-host 'password'
    foreach ($firewall in $firewalList) {
        # Temp is to hold the api key per firewall
        $temp = getapikey($firewall)
        # Add the firewall IP as the first key in hash table and the API key the second value
        $keys.Add($firewall, $temp)
    }
    # Removing passwords from memory
    $user = $null
    $password = $null
}
Measure-Command {
    # Run through the $keys hash table with getenumerator
    foreach ($item in $keys.GetEnumerator()) {
        # Assign the key (IP) to individual variable
        $ip = $item.Key
        # Assign the hash table key's value that has the API key to individual variable
        $key = $item.Value
        # Fetch licenses
        [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><license><fetch></fetch></license></request>&key=$key"
        $result.response.status
        if ($result.response.status -ne "success") { write-host "failed license fetch on $ip" }
        # Check to see if PanOS is up to date
        [xml]$resultsysteminfo = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key"
        $cursoftware = $resultsysteminfo.SelectSingleNode("//sw-version")
        if ($cursoftware.'#text' -ne "$version") {
            # Fetch available software from internet
            [xml]$resultsoft = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><check></check></software></system></request>&key=$key"
            if ($resultsoft.response.status -ne "success") { write-host "failed on Fetch available software $ip" }
            $areyoudownloaded = $resultsoft.SelectSingleNode("//entry[version = ""$version""]") | Select-Object downloaded, current
            if ($areyoudownloaded.current -eq 'no') {
                # Download requested PanOS version
                if ($areyoudownloaded.downloaded -eq 'no') {
                    [xml]$resultdl = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><download><version>$version</version></download></software></system></request>&key=$key"
                    $job = $resultdl.response.result.job
                    isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Downloading PanOS"
                }            
                # Install requested version that's downloaded
                if ($areyoudownloaded.downloaded -eq 'yes') {
                    [xml]$resultinstall = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><install><version>$version</version></install></software></system></request>&key=$key"
                    $job = $resultinstall.response.result.job
                    isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Installing Panos"
                    write-host "$ip has installed $version.  Moving to reboot" 
                }
            }
            if ($resultinstall.response.status -eq "success") {
                #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
                #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
                $confirm = read-host "DO YOU WANT TO REBOOT $ip? Y/N?"
                if ($confirm -eq "Y" -or $confirm -eq "y") {
                    [xml]$resultreboot = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><restart><system></system></restart></request>&key=$key"
                    $resultreboot.response.status
                    while ($areyouinstalled.'#text' -ne "$version") {
                        # Check to see if PanOS is up to date *erroractions are still showing errors*
                        [xml]$resultfin = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key" -ErrorAction SilentlyContinue -WarningAction SilentlyContinue -InformationAction SilentlyContinue
                        $areyouinstalled = $resultfin.SelectSingleNode("//sw-version")
                        write-host "$ip is still booting"
                        if ($areyouinstalled.'#text' -eq "$version") { write-host "$ip has been upgraded" }
                        start-sleep -s 60
                    }
                }
            }
            # Download Apps/Threats
            [xml]$resultapp = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><content><upgrade><download><latest/></download></upgrade></content></request>&key=$key"
            $job = $resultapp.response.result.job
            isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Download apps/threats"
        
            # Install Apps/Threats
            [xml]$resultappins = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><content><upgrade><install><version>latest</version></install></upgrade></content></request>&key=$key"
            $job = $resultappins.response.result.job
            isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Installing apps/threats"
        
            # Download AV
            [xml]$resultavdl = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><anti-virus><upgrade><download><latest/></download></upgrade></anti-virus></request>&key=$key"
            $job = $resultavdl.response.result.job
            isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Downloading AV"
        
            # Install AV
            [xml]$resultavins = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><anti-virus><upgrade><download><latest/></download></upgrade></anti-virus></request>&key=$key"
            $job = $resultavins.response.result.job
            isjobdone -job $job -ip $ip -key $key -whatareyoudoing "Installing AV"
            write-host "$ip has finished"
        }
    }
}


palo alto 8.1.x upgrader rest-api 0.1 alpha.ps1
# Palo Alto Rest API base config upgrade 8.1.x or 9.1.x
# Version: Aplha-0.2 4/22/2020
# Written by Brandon Kelley

# Edit firewallstoupgrade.txt with 1 IP per line
# Connect the devices to the terminal server and mgmt to the switch.
# IP the devices *this alpha version is using DHCP, but it'll probably need to be changed to static,
# Currently there's manual Y/N for the reboot of the firewall after install for saftey reasons
# At this time.

# Future - convert the rest API calls into variables so it's more modular and less code
# Add more error handling and input validation
# Create more functions, so debugging will be easier
# Learn to create parallell proccessing to do multiple firewalls at the same time
# add timer for rebooting

#Ignore self signed certificate calling this script in same dir
$ignorecert = "$PSScriptRoot\ignore self signed certificate.ps1"
&$ignorecert

#text file for IPs you want to configure
$firewalList = (get-content "$PSScriptRoot\firewallstoupgrade.txt") 

#ask user for what version of code you want *need to add 2-3 code jump process via function*
$version = read-host 'version you want to go to'

function getapikey {
    # Get API Key with username and pw prompts
    param ([string]$firewallip)
    [xml]$content = Invoke-RestMethod -uri "https://$firewallip/api/?type=keygen&user=$user&password=$password"
    $content.response.result.key
}
#if there's no key hash table it'll run from here
$keys = $null
if ($keys.values -eq $null) {
    #create keys hash table
    $keys = @{ }
    #get creds
    $user = read-host 'username'
    $password = read-host 'password'
    foreach ($firewall in $firewalList) {
        #temp is to run the api function to get the key per firewall
        $temp = getapikey($firewall)
        # Add the firewall IP as the first key in hash table and the API key as a value
        $keys.Add($firewall, $temp)
    }
    #removing passwords from memory
    $user = $null
    $password = $null
}
Measure-Command {
    # Run through the $keys hash table with getenumerator
    foreach ($item in $keys.GetEnumerator()) {
        #assign the key (IP) to individual variable
        $ip = $item.Key
        #assign the hash table key's value that has the API key to individual variable
        $key = $item.Value
        #invoke license fetch using variables from above
        [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><license><fetch></fetch></license></request>&key=$key"
        $result.response.status
        if ($result.response.status -ne "success") { write-host "failed license fetch on $ip" }
        # if the response isn't success write to screen - expand on this to provide error message or report after it runes
        #Check to see if PanOS is up to date
        [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key"
        $cursoftware = $result.SelectSingleNode("//sw-version")
        if ($cursoftware -ne "$version") {
            #Fetch available software from internet
            [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><check></check></software></system></request>&key=$key"
            if ($result.response.status -ne "success") { write-host "failed on Fetch available software $ip" }
            $areYouUpgraded = $result.SelectSingleNode("//entry[version = ""$version""]") | Select-Object downloaded, uploaded
            if ($areYouUpgraded.downloaded -eq 'no' -and $areYouUpgraded.uploaded -eq 'no') {
                #download $version pan os *turn that into a variable*
                [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><download><version>$version</version></download></software></system></request>&key=$key"
                $job = $result.response.result.job
                $stoploop = $result.SelectSingleNode("//job") | Select-Object status, result
                #Loop to check on PanOS downloading 
                while ($stoploop.result -ne "OK" -and $stoploop.status -ne "FAIL") {
                    [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
                    write-host $result.response.result.job.progress "% Downloading PanOS complete"
                    $stoploop = $result.SelectSingleNode("//job") | Select-Object status, result #change the status to install one
                    start-sleep -s 15
                }
            }
            Else {
                [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><install><version>$version</version></install></software></system></request>&key=$key"
                $job = $result.response.result.job
               while ($stoploop.result -ne "OK" -and $stoploop.status -ne "FIN") {
                    [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><jobs><id>$job</id></jobs></show>&key=$key"
                    write-host $result.response.result.job.progress "% complete Installing PanOS"
                    $stoploop = $result.SelectSingleNode("//job") | Select-Object status, result
                    start-sleep -s 15
                }
            }
        }
        #verify panOS insatlled
        [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><system><software><check></check></software></system></request>&key=$key"
        #add another condition to check for uploaded also
        $areyouinstalled = $result.SelectSingleNode("//entry[version = ""$version""]") | Select-Object current, downloaded
        if ($areyouinstalled.current -eq "no" -and $areyouinstalled.downloaded -eq 'yes') {
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            #!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!!DANGER OPERATIONAL COMMAND TO REBOOT!!!!!!!!!!!!!
            $confirm = read-host "DO YOU WANT TO REBOOT $ip? Y/N?"
            if ($confirm -eq "Y" -or $confirm -eq "y") {
                [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<request><restart><system></system></restart></request>&key=$key"
                $result
                while ($cursoftware -ne "$version") {
                    #Check to see if PanOS is up to date *erroraction isn't working right fix it*
                    [xml]$result = Invoke-RestMethod -method Get -uri "https://$ip/api/?type=op&cmd=<show><system><info></info></system></show>&key=$key" -ErrorAction SilentlyContinue
                    $cursoftware = $result.SelectSingleNode("//sw-version")
                    write-host "$ip is still booting"
                    start-sleep -s 15
                    if ($cursoftware -eq "$version" -and $result.response.result -ne "Command succeeded with no output") { write-host "$ip has been upgraded" }
                }
            }
        }
        #learn parallell processing in powershell to make it do all firewalls in list at same time
    }
    #Reset variables for next Upgrade
    $cursoftware = $null
    $areyouinstalled = $null
    $areYouUpgraded = $null
    $job = $null
    $result = $null
    #finish downloading/installing app/threats/av
    #add error handling
}
